### PR TITLE
Document how testing works 

### PR DESCRIPTION
**Pull request by lmorchard**
issue #2977

### CODE REVIEW

**Comment by lmorchard** [comment]
Rather than bury this in a google doc, I'm opening this up as a draft PR with a markdown file under `docs/`. Going to just start pushing updates to it as I go. This is living on a branch in the origin repo so other folks can also push changes to it as necessary, otherwise feel free to comment / review / nitpick at any time

File: docs/how-testing-works.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+Logic for running or skipping build for a given module is implemented here: `packages/test.list` is expected to contain a list output from `modules-to-test.js` - if the current module's name is missing and the list isn't `all`, this script will exit immediately.
+
+Parameters are just `$1` for the name of the module being tested.
+
+The file `packages/version.json` is expected to have been generated earlier - i.e. from the `install` job in CircleCI. This file is copied into the package.
+
+For modules where `Dockerfile` is present, a container tagged `{MODULE}:build` is built from `Dockerfile`.
+
+For modules where `Dockerfile-build` is present, the container is built from that file instead.
```
Comment (shane-tomlinson) [inline docs/how-testing-works.md]: Not for this PR, but why are there both Dockerfile and Dockerfile-build files?

File: docs/how-testing-works.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+#### Workflows
+
+##### test
+
+##### deploy-tag
+
+### TeamCity
+
+TBD: what's TeamCity? who administers it? when does it run tests? what tests does it run?
```
Comment (shane-tomlinson) [inline docs/how-testing-works.md]: I can help with this section whenever I get some time.

File: docs/how-testing-works.md (hunk 1 above)
Comment (lmorchard) [reply docs/how-testing-works.md]: Good question! 🤷‍♂

File: docs/how-testing-works.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+For master branch and other non-PR commits, the output is always `all`.
+
+For pull requests, the script fetches and parses the diff for the current PR. The URL for the diff is constructed from CircleCI env vars.
+
+If there's an error while fetching or parsing the diff, `all` is output. Otherwise, the script checks the path of each file changed to build a list of modules for testing.
+
+Some modules are depend on other modules. Within `package.json` at the root of the FxA project, there is an `fxa.moduleDependencies` object. This maps named modules to their dependencies.
+
+If the diff contains changes to any of the dependencies, the dependent module is also added to the list. So, for example, `fxa-content-server` lists `fxa-auth-server` as a dependency - thus, a change to `fxa-auth-server` causes `fxa-content-server` to also be included in the list.
```
[the remark this answers is not in the document]
Comment (lmorchard) [reply docs/how-testing-works.md]: Yeah, I could `s/to also be included/to be automatically included/` to make that clearer

File: docs/how-testing-works.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+CircleCI is a service that automatically runs FxA tests in parallel jobs for every Pull Request and merge to `master` branch on GitHub. Scripts and configuration for CircleCI mostly lives in the `.circleci` directory.
+
+The file `.circleci/config.yml` contains general configuration, job definitions, and workflows orchestrating all the jobs.
+
+#### Important scripts
+
+##### .circleci/modules-to-test.js
+
+Running all tests across all modules / packages in FxA can be time-comsuming. Time can be saved by skipping tests that we know are unrelated to changes in some given commit. Applying that logic, this script outputs a name of individual packages that should be tested - or `all` if it decides that everything needs to be run.
```
Comment (6a68) [inline docs/how-testing-works.md]: typo: "time-consuming"

**Comment by 6a68** [comment]
nice work @lmorchard! This looks great so far :+1: :+1:

File: docs/how-testing-works.md (hunk 1 above)
Comment (dannycoates) [reply docs/how-testing-works.md]: No good reason. Just an quirk of the previous multi-repos

**Comment by lmorchard** [comment]
Rebased against master and made another chunk of progress trying to describe how the general CircleCI process works.

**Comment by clouserw** [comment]
wow, there are a lot of things going on here.  Thanks for digging through the details here.  I think it's looking great.  My only comment would be to use https://github.com/mozilla/ecosystem-platform once you get it to a state that's ready to commit.  I think it would fit as a "deep dive"

**Comment by lmorchard** [comment]
Closing this draft for now. Going to re-open as a PR over on ecosystem-docs after some edits

**Pull request closed without merging**